flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java

```diff
@@ -104,8 +104,12 @@ protected JobExceptionsInfo handleRequest(
 
     private static JobExceptionsInfo createJobExceptionsInfo(
             ExecutionGraphInfo executionGraphInfo, int exceptionToReportMaxSize) {
+        final List<TaskFailureHistoryEntry> rootCauseFirstExceptionHistory =
+                executionGraphInfo.getExceptionHistory();
+        Collections.reverse(rootCauseFirstExceptionHistory);
```

Comment:
I feel like such business should be handled in the UI.

Reply:
You have a point. It's just that reversing the order makes it easier to select the root cause which is the last element.

Reply:
This is resolved after introducing the `EvictingBoundedList` internally.